keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,548 @@
+   * a feature is disabled if the feature is proposed to be `false` by any member
+   * a feature is enabled only if the feature is proposed to be `true` by all members
+   * a feature is disabled if it is missing from the proposal of any member.
+1. The leader will send `ClusterFeatureGateSetRequest` through raft to set the final values of the cluster feature gate to all members. 
+
+One drawback of this approach is that there is a period before the leader receives the proposals from all members when no feature would be available. The duration of this period is expected to be `O(time to set ClusterVersion)`. We think that is acceptable because if the time to set `ClusterVersion` is exceptionally long, the `ClusterVersion` would stay at the fallback version `3.0`, for which a lot of etcd capabilities would not be available anyway. It is better to wait than risking splitting minds about what the default values should be when relying on falling back to default values of the `ClusterVersion`.
+
+#### Restart a Member
+
+Fig. b above shows how the consensus of feature gates is reached when a member of an existing cluster restarts:
+1. When the member bootstraps, it will load the saved `ClusterVersion` and `ClusterFeatureGates` from the backend. Since the `ClusterVersion` should be `<=` the server binary version when downgrade is not enabled, the `ClusterFeatureGates` should also be compatible with the server binary version (we do not allow feature backport).
+1. After bootstrapping, the new server will use the new `--cluster-feature-gates` flag to update its new `proposed_feature_gates` through `member.Attributes` as part of the server [`Start()`](https://github.com/etcd-io/etcd/blob/67743348dcd2deb1819fded2336749449ebf95d2/server/etcdserver/server.go#L532) process.
+1. The leader will received the updated `member.Attributes.proposed_feature_gates` message from the restarted member, and update the cluster feature gates for the whole cluster.
+
+#### Remove a New Member
+
+Fig. c above shows how the consensus of feature gates is reached when removing a member to an existing cluster:
+1. When the leader apply the config change to remove a member, it will rerun the `UpdateClusterFeatureGate` logic, update the cluster feature gate based on the `proposed_feature_gates` of the remaining members, and send the updated cluster feature gate values to the whole cluster.
+
+#### Add a New Member
+
+Fig. d above shows how the consensus of feature gates is reached when adding a new member to an existing cluster:
+1. When the cluster applies the ConfChange request to add a member, the leader will call `UpdateClusterFeatureGate` which will reset the cluster feature gate because `proposed_feature_gates` of the new member is missing.
```

Comment:
Could we exclude member until it's bootstrapped to avoid flip-flops?

Reply:
sg.